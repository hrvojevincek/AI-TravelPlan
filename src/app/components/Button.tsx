"use client";
import React from "react";
import { signIn, signOut, useSession } from "next-auth/react";
import { ArrowRightOnRectangleIcon } from "@heroicons/react/20/solid";

const Button = ({
  className,
  ...props
}: React.DetailedHTMLProps<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
>) => {
  const { data: session } = useSession();

  const popupCenter = (url: string, title: string) => {
    const dualScreenLeft = window.screenLeft ?? window.screenX;
    const dualScreenTop = window.screenTop ?? window.screenY;

    const width =
      window.innerWidth ?? document.documentElement.clientWidth ?? screen.width;

    const height =
      window.innerHeight ??
      document.documentElement.clientHeight ??
      screen.height;

    const systemZoom = width / window.screen.availWidth;

    const left = (width - 500) / 2 / systemZoom + dualScreenLeft;
    const top = (height - 550) / 2 / systemZoom + dualScreenTop;

    const newWindow = window.open(
      url,
      title,
      `width=${500 / systemZoom},height=${
        550 / systemZoom
      },top=${top},left=${left}`
    );

    newWindow?.focus();
  };

  if (session && session.user) {
    return (
      <>
        {/* <div className="flex gap-4"> */}
        {/* <p className="text-sky-600">{session.user.name}</p> */}
        <button onClick={() => signOut()} className={`${className}`} {...props}>
          <ArrowRightOnRectangleIcon className="inline w-4 h-4 mr-1" />
          <span>Sign Out</span>
        </button>
        {/* </div> */}
      </>
    );
  }
  return (
    <button
      onClick={() => popupCenter("/google-signin", "Sample Sign In")}
      className={`${className}`}
      {...props}
    >
      Sign In
    </button>
  );
};

export default Button;
