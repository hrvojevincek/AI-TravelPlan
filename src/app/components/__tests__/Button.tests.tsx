import { fireEvent, render } from "@testing-library/react";
import Button from "../Button";
import { useSession, signIn, signOut } from "next-auth/react";
jest.mock("next-auth/react");
global.window.open = jest.fn();

describe("signed out", () => {
  beforeAll(() => {
    (useSession as jest.Mock).mockReturnValue({ data: {} });
  });

  it("renders signed in button", () => {
    const { container } = render(<Button />);
    expect(container).toMatchSnapshot();
  });

  it("calls sign in when clicked", () => {
    const { getByText } = render(<Button />);
    fireEvent.click(getByText("Sign In"));
    expect(global.window.open).toHaveBeenCalledTimes(1);
    expect(global.window.open).toHaveBeenCalledWith(
      "/google-signin",
      expect.anything(),
      expect.anything()
    );
  });

  afterAll(() => {
    (useSession as jest.Mock).mockReset();
  });
});

describe("signed in", () => {
  beforeAll(() => {
    (useSession as jest.Mock).mockReturnValue({
      data: {
        user: {
          name: "John",
        },
      },
    });
  });
  it("renders singed out button", () => {
    const { container } = render(<Button />);
    expect(container).toMatchSnapshot();
  });

  it("calls sign out when clicked", () => {
    const { getByText } = render(<Button />);
    fireEvent.click(getByText("Sign Out"));
    expect(signOut).toHaveBeenCalledTimes(1);
  });

  afterAll(() => {
    (useSession as jest.Mock).mockReset();
  });
});
