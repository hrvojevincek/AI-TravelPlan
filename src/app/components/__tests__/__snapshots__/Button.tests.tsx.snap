// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signed in renders singed out button 1`] = `
<div>
  <div
    class="flex gap-4 ml-auto"
  >
    <p
      class="text-sky-600"
    >
      John
    </p>
    <button
      class="text-red-600"
    >
      Sign Out
    </button>
  </div>
</div>
`;

exports[`signed out renders signed in button 1`] = `
<div>
  <button
    class="text-green-600 ml-auto"
  >
    Sign In
  </button>
</div>
`;
